src/tracks/playabletrack/wavetrack/ui/SampleHandle.cpp

```diff
@@ -93,9 +99,17 @@ namespace {
    {
       // Require more than 3 pixels per sample
       const auto xx = std::max<ZoomInfo::int64>(0, viewInfo.TimeToPosition(time));
+      const auto& clips = wt->GetClips();
+      const auto clipIt =
+         std::find_if(clips.begin(), clips.end(), [time](const auto& clip) {
+            return clip->WithinPlayRegion(time);
+         });
       ZoomInfo::Intervals intervals;
-      const double rate = wt->GetRate();
-      viewInfo.FindIntervals(rate, intervals, width);
+      const double rate =
+         clipIt == clips.end() ?
+            wt->GetRate() :
+            (*clipIt)->GetRate() / (*clipIt)->GetStretchRatio();
+      viewInfo.FindIntervals(intervals, width);
```

Comment:
SampleHandle.cpp is the first part of this pull request that I am studying more deeply.

I see how this function, used in hit testing, now changes rate when inside a clip.

But there is still a use of `WaveTrack::GetFloats` in `SampleHandle::HitTest.`

Should we eliminate all uses of `GetFloats` in here?

Should hit test fail if the cursor is not inside the x coordinate boundaries of a clip?  (or within some pixel width tolerande or the boundaries?)

Reply:
I also see `GetEnvelopeAtTime` is used, but that calls `GetClipAtTime`, so if the latter hit-tests correctly for the x coordinate, so does the former.

Reply:
`Envelope::GetValue` is also called, passed a sample duration.

Maybe that use of rate needs to be corrected for the clip, and maybe it is unimportant when getting just one value.

Reply:
Nice that you found that `GetValue` use and the potential for confusion the sample-step argument has.
AFAICS in the implementation, step size only becomes relevant for queries of more than one value. This step size defaults to 0. With more time I'd consider removing that optional argument and hard-code it to zero when it calls `GetValues` underneath. But here I'll just use the default value instead of the sample period.

Reply:
> Should we eliminate all uses of GetFloats in here?

It's preferable ; using `GetFloatAt` instead. Thank you.